### PR TITLE
Component import

### PR DESCRIPTION
**Pull request by omarabed15**
This PR allows you to import a component into the Component Library with properly-formatted OSCAL JSON. On /controls/components, use the "Import OSCAL Component" button to bring up the dialog.

### CODE REVIEW

**Comment by davidpofo** [comment]
Just needs some tests, other than that lgtm.

**Comment by omarabed15** [comment]
> Just needs some tests, other than that lgtm.

This should be good to go now. Tests are written and pass, and they check for both successful and unsuccessful scenarios.

**Review by davidpofo — APPROVED** [review]
lgtm

File: controls/views.py [hunk 1]
```diff
@@ -329,6 +336,119 @@ def as_yaml(self):
...
+        @returns: Statement object if created, None otherwise
+        """
+
+        existing_statement_uuids = Statement.objects.filter(uuid=impl_stmnt_json['uuid']).count()
+        if existing_statement_uuids > 0:
+            messages.add_message(request, messages.ERROR, f"Statement with this uuid already exists. Skipping...")
+            return None
+        try:
+            new_statement = Statement.objects.create(
+                sid=impl_stmnt_json['implemented-requirements'][0]['control-id'] if 'control-id' in impl_stmnt_json['implemented-requirements'][0] else None,
```
[the remark this answers is not in the document]
Comment (omarabed15) [reply controls/views.py:448]: The NIST OSCAL JSON schema specifies the Implemented Requirements (i.e. Implemented Controls) of a Statement as an array, but requires that there is at least one. To my knowledge, GovReady only supports one Control per Implementation Statement (at least, through the Component Library), so this code is pulling the Control from the first Implemented Requirement / Control in the array. Down the road, we may be able to support multiple Controls linked from a single Statement.

File: controls/views.py [hunk 2]
```diff
@@ -329,6 +336,119 @@ def as_yaml(self):
...
+        """Creates a component from a JSON dict
+
+        @type component_uuid: str
+        @param component_uuid: UUID of imported component
+        @type component_json: dict
+        @param component_json: Component attributes from JSON object
+        @rtype: Element
+        @returns: Element object if created, None otherwise
+        """
+        existing_element_uuids = Element.objects.filter(uuid=component_uuid).count()
```
[the remark this answers is not in the document]
Comment (omarabed15) [reply controls/views.py]: This should be resolved for both Elements and Statements. In the Statements, the UUID is the value for the "uuid" key, but for the Element, the UUID is the key itself, so I passed that value into the create_component function as the first parameter.

File: controls/views.py [hunk 3]
```diff
@@ -329,6 +338,129 @@ def as_yaml(self):
...
+        else:
+            messages.add_message(request, messages.ERROR, f"Invalid OSCAL. Component(s) not created.")
+            return False
+
+    def validate_oscal_json(self, oscal_json):
+        """Validates the JSON object is valid OSCAL format"""
+
+        project_root = os.path.abspath(os.path.dirname(__name__))
+        oscal_schema_path = os.path.join(project_root, "schemas", "oscal_component_schema.json")
+        with open(oscal_schema_path, "r") as schema_content:
```
[the remark this answers is not in the document]
Comment (omarabed15) [reply controls/views.py:371]: It's restricted on the frontend by the file select field, but agreed, it's a good thing to check. What ends up happening here is actually that the file doesn't get used, but its contents get poured into the textarea to allow the user to adjust them before importing, if desired. So _technically_ any file type can be used, and then adjusted in the textarea, but for simplicity's sake, the file select is restricted to .json on the frontend.

File: controls/views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (omarabed15) [reply controls/views.py]: This should be fixed now. I'm looping through the statements, and checking for the existence of the control in the catalog before creating the statement.

File: templates/components/component_library.html [hunk 4]
```diff
@@ -22,35 +22,34 @@
...
-        <div id="" class="col-xs-8 col-sm-8 col-md-8 col-lg-8 col-xl-8">
-            {{ component.description }}
-        </div>
-
-    </div>
-    {% endfor %}
 
-      </form>
+  <div id="tab-content" class="row rows-header" style="">
+    <div id="" class="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-xl-3"><span style="font-size: 14pt; font-weight: bold;">Available Components</div>
```
Comment (gregelin) [inline templates/components/component_library.html:37]: We need to move the inline styles to `govready-q.css`.

File: templates/systems/components_selected.html [hunk 5]
```diff
@@ -28,13 +28,6 @@ <h2 class="control-heading">
             <a href="{{ project.get_absolute_url }}" class="systems-link">{{ system.root_element.name }}</a> > Selected Components
           </h2>
       </div>
-      <!-- Control Lookup-->
-      <div class="systems-control-lookup">
```
[the remark this answers is not in the document]
Comment (gregelin) [reply templates/systems/components_selected.html:32]: We can probably delete commented out form block.

**Review by davidpofo — APPROVED** [review]

**Pull request merged**